client/caps.go

```diff
@@ -45,20 +46,57 @@ type Capability struct {
 
 // Capabilities returns the capabilities currently available for snaps to consume.
 func (client *Client) Capabilities() (map[string]Capability, error) {
+	errPrefix := "cannot obtain capabilities"
 	var rsp response
 	if err := client.do("GET", "/1.0/capabilities", nil, &rsp); err != nil {
 		return nil, err
 	}
-	if rsp.Type == "error" {
-		// TODO: handle structured errors
-		return nil, fmt.Errorf("cannot obtain capabilities: %s", rsp.Status)
+	switch rsp.Type {
+	case "error":
+		return nil, rsp.processErrorResponse()
+	case "sync":
+		var resultOk map[string]map[string]Capability
+		if err := json.Unmarshal(rsp.Result, &resultOk); err != nil {
+			return nil, fmt.Errorf("%s: failed to unmarshal response: %v", errPrefix, err)
+		}
+		return resultOk["capabilities"], nil
+	default:
+		return nil, fmt.Errorf("%s: expected sync response, got %s", errPrefix, rsp.Type)
 	}
-	if rsp.Type != "sync" {
-		return nil, fmt.Errorf("cannot obtain capabilities: expected sync response, got %s", rsp.Type)
+}
+
+// AddCapability adds one capability to the system
+func (client *Client) AddCapability(c *Capability) error {
+	b, err := json.Marshal(c)
+	if err != nil {
+		return err
+	}
+	var rsp response
+	if err := client.do("POST", "/1.0/capabilities", bytes.NewReader(b), &rsp); err != nil {
+		return err
+	}
+	switch rsp.Type {
+	case "error":
+		return rsp.processErrorResponse()
+	case "sync":
+		return nil
+	default:
+		return fmt.Errorf("cannot obtain capabilities: expected sync response, got %s", rsp.Type)
+	}
+}
+
+// RemoveCapability removes one capability from the system
+func (client *Client) RemoveCapability(name string) error {
+	var rsp response
+	if err := client.do("DELETE", fmt.Sprintf("/1.0/capabilities/%s", name), nil, &rsp); err != nil {
+		return err
 	}
-	var result map[string]map[string]Capability
-	if err := json.Unmarshal(rsp.Result, &result); err != nil {
-		return nil, fmt.Errorf("cannot obtain capabilities: failed to unmarshal response: %v", err)
+	switch rsp.Type {
```

Comment:
This can use the same `if err := rsp.err(); err != nil` trick mentioned in the underlying branch.